pkg/chunk/purger/blocks_purger.go

```diff
@@ -0,0 +1,234 @@
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+	}
+
+	endParam := params.Get("end")
+	endTime := int64(model.Now())
+
+	if endParam != "" {
+		endTime, err = util.ParseTime(endParam)
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		if endTime > int64(model.Now()) {
+			http.Error(w, "deletes in future not allowed", http.StatusBadRequest)
+			return
+		}
+	}
+
+	if startTime > endTime {
+		http.Error(w, "start time can't be greater than end time", http.StatusBadRequest)
+		return
+	}
+
+	tManager := cortex_tsdb.NewTombstoneManager(api.bucketClient, userID, api.cfgProvider, api.logger)
+
+	requestID := getTombstoneHash(startTime, endTime, match)
+	// Since the request id is based on a hash of the parameters, there is a possibility that a tombstone could already exist for it
+	// if the request was previously cancelled, we need to remove the cancelled tombstone before adding the pending one
+	if err := tManager.RemoveCancelledStateIfExists(ctx, requestID); err != nil {
+		level.Error(util_log.Logger).Log("msg", "removing cancelled tombstone state if it exists", "err", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	prevT, err := tManager.GetTombstoneByIDForUser(ctx, requestID)
+	if err != nil {
+		level.Error(util_log.Logger).Log("msg", "error getting delete request by id", "err", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	if prevT != nil {
+		http.Error(w, "delete request tombstone with same information already exists", http.StatusBadRequest)
+		return
+	}
+
+	curTime := time.Now().Unix() * 1000
+	t := cortex_tsdb.NewTombstone(userID, curTime, curTime, startTime, endTime, match, requestID, cortex_tsdb.StatePending)
+
+	if err = tManager.WriteTombstone(ctx, t); err != nil {
+		level.Error(util_log.Logger).Log("msg", "error adding delete request to the object store", "err", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	w.WriteHeader(http.StatusNoContent)
+}
+
+func (api *BlocksPurgerAPI) GetAllDeleteRequestsHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	userID, err := tenant.TenantID(ctx)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusUnauthorized)
+		return
+	}
+	tManager := cortex_tsdb.NewTombstoneManager(api.bucketClient, userID, api.cfgProvider, api.logger)
+	deleteRequests, err := tManager.GetAllTombstonesForUser(ctx)
+	if err != nil {
+		level.Error(util_log.Logger).Log("msg", "error getting delete requests from the block store", "err", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	if err := json.NewEncoder(w).Encode(deleteRequests); err != nil {
+		level.Error(util_log.Logger).Log("msg", "error marshalling response", "err", err)
+		http.Error(w, fmt.Sprintf("Error marshalling response: %v", err), http.StatusInternalServerError)
+	}
+
+	w.WriteHeader(http.StatusOK)
```

Comment:
According to https://pkg.go.dev/net/http#ResponseWriter since you  called `json.NewEncoder(w)`, which I assume calls `ResponseWriter.Write()`, then status code of 200 is automatically written for you together with "guessed" content-type. So you don't really need to set the status at this line. 

However, I think we should set the content type and the status header ourselves for clarity. 

Also, I think the if block on line 155 is missing a return statement right? I would expect `http.Error()` writes that the status header  to 500 already, and this `w.WriteHeader(http.StatusOk)` may have undefined behaviour?

Reply:
Makes sense, should be fixed now.